python/taichi/lang/util.py

```diff
@@ -226,4 +228,69 @@ def wrapped(*args, **kwargs):
     return wrapped
 
 
+# The builtin `warnings` module is unreliable since it may be suppressed
+# by other packages such as IPython.
+def warning(msg, warning_type=UserWarning, stacklevel=1):
+    """Print warning message
+
+    Args:
+        msg (str): massage to print.
+        warning_type (builtin warning type):  type of warning.
+        stacklevel (int): warning stack level from the caller.
+    """
+    s = traceback.extract_stack()[:-stacklevel]
+    raw = ''.join(traceback.format_list(s))
+    print(Fore.YELLOW + Style.BRIGHT, end='')
+    print(f'{warning_type.__name__}: {msg}')
+    print(f'\n{raw}')
+    print(Style.RESET_ALL, end='')
+
+
+def deprecated(old, new, warning_type=DeprecationWarning):
+    """Mark an API as deprecated.
+
+    Args:
+        old (str): old method.
+        new (str): new method.
+        warning_type (builtin warning type): type of warning.
+
+    Example::
+
+        >>> @deprecated('ti.sqr(x)', 'x**2')
+        >>> def sqr(x):
+        >>>     return x**2
+
+    Returns:
+        Decorated fuction with warning message
+    """
+    def decorator(foo):
+        @functools.wraps(foo)
+        def wrapped(*args, **kwargs):
+            msg = f'{old} is deprecated. Please use {new} instead.'
+            warning(msg, warning_type, stacklevel=2)
+            return foo(*args, **kwargs)
+
+        return wrapped
+
+    return decorator
+
+
+def obsolete(old, new):
```

Comment:
Can we do this in `__getattr__` as well?

Reply:
Yes but actually I think this can be removed. We don't need so many levels for deprecating things for now and this is only used in one place one year ago.